syntax = "proto3";
package pt.ulisboa.tecnico.nameserver.contract;

message RegisterRequest {
  string serviceName = 1;
  string qualifier = 2;
  string target = 3;
}

message RegisterResponse {
  string error = 1;
}

message LookupRequest {
  string serviceName = 1;
  string qualifier = 2;
}

message ServerInfo {
  string qualifier = 1;
  string target = 2;
}
message LookupResponse {
  repeated ServerInfo servers = 1;
}

message GenerateIdRequest {
}

message GenerateIdResponse {
  int32 clientId = 1;
  int32 serverNumber = 2;
}

message DeleteRequest {
  string serviceName = 1;
  string target = 2;
}

message DeleteResponse {
  string error = 1;
}

service TupleSpacesNameServer {
  rpc Register(RegisterRequest) returns (RegisterResponse);
  rpc Lookup(LookupRequest) returns (LookupResponse);
  rpc GenerateId(GenerateIdRequest) returns (GenerateIdResponse);
  rpc Delete(DeleteRequest) returns (DeleteResponse);
}
